File: apps/host/internal/githubhost/mapping_test.go
package githubhost

import (
	"errors"
	"testing"

	pb "armadra.local/host/gen/armadra/v1"
)

func labelGroup(id, label string, couples pb.GithubIssueState) *pb.GithubStatusGroup {
	return &pb.GithubStatusGroup{Id: id, Title: id, Label: label, CouplesIssueState: couples}
}

// Both coupling directions together can describe a loop: entering one group
// would set a state that selects another group, which sets a state that selects
// the first. That would oscillate against the remote forever, so it is refused
// before anything is stored.
func TestMappingCyclesAreRefused(t *testing.T) {
	cyclic := &pb.GithubStatusMapping{
		Source: pb.GithubStatusSource_GITHUB_STATUS_SOURCE_LABEL,
		Groups: []*pb.GithubStatusGroup{
			labelGroup("done", "status/done", pb.GithubIssueState_GITHUB_ISSUE_STATE_CLOSED),
			labelGroup("todo", "status/todo", pb.GithubIssueState_GITHUB_ISSUE_STATE_OPEN),
		},
		StateGroups: []*pb.GithubStateCoupling{
			{State: pb.GithubIssueState_GITHUB_ISSUE_STATE_CLOSED, GroupId: "todo"},
			{State: pb.GithubIssueState_GITHUB_ISSUE_STATE_OPEN, GroupId: "done"},
		},
	}
	reason, err := ValidateMapping(cyclic)
	if !errors.Is(err, ErrInvalid) || reason != "MAPPING_CYCLE" {
		t.Fatalf("a cyclic mapping was accepted (%s, %v)", reason, err)
	}
	// A group that maps back to itself is a fixed point, not a loop: entering
	// "done" closes the Issue, and a closed Issue is in "done". That is
	// consistent and must stay configurable.
	fixed := &pb.GithubStatusMapping{
		Source: pb.GithubStatusSource_GITHUB_STATUS_SOURCE_LABEL,
		Groups: []*pb.GithubStatusGroup{
			labelGroup("done", "status/done", pb.GithubIssueState_GITHUB_ISSUE_STATE_CLOSED),
			labelGroup("todo", "status/todo", pb.GithubIssueState_GITHUB_ISSUE_STATE_UNSPECIFIED),
		},
		StateGroups: []*pb.GithubStateCoupling{{State: pb.GithubIssueState_GITHUB_ISSUE_STATE_CLOSED, GroupId: "done"}},
	}
	if reason, err = ValidateMapping(fixed); err != nil {
		t.Fatalf("a fixed point was refused (%s, %v)", reason, err)
	}
	// A longer cycle through three groups must be caught too.
	long := &pb.GithubStatusMapping{
		Source: pb.GithubStatusSource_GITHUB_STATUS_SOURCE_LABEL,
		Groups: []*pb.GithubStatusGroup{
			labelGroup("a", "status/a", pb.GithubIssueState_GITHUB_ISSUE_STATE_OPEN),
			labelGroup("b", "status/b", pb.GithubIssueState_GITHUB_ISSUE_STATE_CLOSED),
			labelGroup("c", "status/c", pb.GithubIssueState_GITHUB_ISSUE_STATE_UNSPECIFIED),
		},
		StateGroups: []*pb.GithubStateCoupling{
			{State: pb.GithubIssueState_GITHUB_ISSUE_STATE_OPEN, GroupId: "b"},
			{State: pb.GithubIssueState_GITHUB_ISSUE_STATE_CLOSED, GroupId: "a"},
		},
	}
	if reason, err = ValidateMapping(long); !errors.Is(err, ErrInvalid) || reason != "MAPPING_CYCLE" {
		t.Fatalf("a two-step cycle was accepted (%s, %v)", reason, err)
	}
}

// Each configuration mistake gets its own stable reason, because a user has to
// find and fix it rather than retry.
func TestMappingValidationNamesTheMistake(t *testing.T) {
	label := func(groups ...*pb.GithubStatusGroup) *pb.GithubStatusMapping {
		return &pb.GithubStatusMapping{Source: pb.GithubStatusSource_GITHUB_STATUS_SOURCE_LABEL, Groups: groups}
	}
	cases := map[string]*pb.GithubStatusMapping{
		"MAPPING_REQUIRED":      nil,
		"SOURCE_UNSPECIFIED":    {Source: pb.GithubStatusSource_GITHUB_STATUS_SOURCE_UNSPECIFIED},
		"GROUPS_REQUIRED":       label(),
		"GROUP_LABEL_DUPLICATE": label(labelGroup("a", "same", 0), labelGroup("b", "same", 0)),
		"GROUP_ID_INVALID":      label(labelGroup("a", "x", 0), labelGroup("a", "y", 0)),
		"GROUP_LABEL_INVALID":   label(&pb.GithubStatusGroup{Id: "a", Title: "a"}),
		"LABEL_SOURCE_HAS_PROJECT": {
			Source:    pb.GithubStatusSource_GITHUB_STATUS_SOURCE_LABEL,
			ProjectId: "PVT_1",
			Groups:    []*pb.GithubStatusGroup{labelGroup("a", "x", 0)},
		},
		"PROJECT_IDS_REQUIRED": {
			Source: pb.GithubStatusSource_GITHUB_STATUS_SOURCE_PROJECT_FIELD,
			Groups: []*pb.GithubStatusGroup{{Id: "a", Title: "a", ProjectOptionId: "opt"}},
		},
		"SOURCE_NONE_HAS_GROUPS": {
			Source: pb.GithubStatusSource_GITHUB_STATUS_SOURCE_NONE,
			Groups: []*pb.GithubStatusGroup{labelGroup("a", "x", 0)},
		},
		"COUPLING_GROUP_UNKNOWN": {
			Source:      pb.GithubStatusSource_GITHUB_STATUS_SOURCE_LABEL,
			Groups:      []*pb.GithubStatusGroup{labelGroup("a", "x", 0)},
			StateGroups: []*pb.GithubStateCoupling{{State: pb.GithubIssueState_GITHUB_ISSUE_STATE_OPEN, GroupId: "missing"}},
		},
		"COUPLING_STATE_DUPLICATE": {
			Source: pb.GithubStatusSource_GITHUB_STATUS_SOURCE_LABEL,
			Groups: []*pb.GithubStatusGroup{labelGroup("a", "x", 0)},
			StateGroups: []*pb.GithubStateCoupling{
				{State: pb.GithubIssueState_GITHUB_ISSUE_STATE_OPEN, GroupId: "a"},
				{State: pb.GithubIssueState_GITHUB_ISSUE_STATE_OPEN, GroupId: "a"},
			},
		},
	}
	for expected, mapping := range cases {
		reason, err := ValidateMapping(mapping)
		if !errors.Is(err, ErrInvalid) || reason != expected {
			t.Fatalf("expected %s, got %s (%v)", expected, reason, err)
		}
	}
	// A configuration with no source at all is valid: not configured is a
	// normal state, and the panel shows Issues ungrouped.
	if reason, err := ValidateMapping(&pb.GithubStatusMapping{Source: pb.GithubStatusSource_GITHUB_STATUS_SOURCE_NONE}); err != nil {
		t.Fatalf("an unconfigured mapping was refused (%s, %v)", reason, err)
	}
}

// Two configured labels on one Issue make its group ambiguous. The panel is
// told that, rather than being handed an arbitrary winner.
func TestConflictingLabelsAreReportedRatherThanResolved(t *testing.T) {
	mapping := &pb.GithubStatusMapping{
		Source: pb.GithubStatusSource_GITHUB_STATUS_SOURCE_LABEL,
		Groups: []*pb.GithubStatusGroup{labelGroup("todo", "status/todo", 0), labelGroup("done", "status/done", 0)},
	}
	both := &pb.GithubIssue{Number: 1, Labels: []*pb.GithubLabel{{Name: "status/todo"}, {Name: "status/done"}, {Name: "bug"}}}
	one := &pb.GithubIssue{Number: 2, Labels: []*pb.GithubLabel{{Name: "status/done"}, {Name: "bug"}}}
	none := &pb.GithubIssue{Number: 3, Labels: []*pb.GithubLabel{{Name: "bug"}}}
	applyLabelGroups(mapping, []*pb.GithubIssue{both, one, none})
	if !both.StatusConflict || both.StatusGroupId != "" {
		t.Fatalf("an ambiguous Issue reported group %q, conflict %v", both.StatusGroupId, both.StatusConflict)
	}
	if one.StatusGroupId != "done" || one.StatusConflict {
		t.Fatalf("a single match reported %q", one.StatusGroupId)
	}
	if none.StatusGroupId != "" || none.StatusConflict {
		t.Fatal("an unmapped Issue must be unmapped, not conflicted")
	}
}

File: apps/desktop/src/core/github/mapping.test.ts
import { describe, expect, it } from "vitest";

import { GithubIssueState, GithubStatusSource } from "./types";
import {
  GithubIssueSchema,
  GithubLabelSchema,
  GithubStateCouplingSchema,
  GithubStatusGroupSchema,
  GithubStatusMappingSchema,
} from "./schema";
import { create } from "../contract/message";

import { applyLabelGroups, applyProjectGroups, checkMapping } from "./mapping";

function group(
  id: string,
  label: string,
  couples: GithubIssueState = GithubIssueState.UNSPECIFIED,
) {
  return create(GithubStatusGroupSchema, {
    id,
    title: id,
    label,
    couplesIssueState: couples,
  });
}

function labelMapping(...groups: ReturnType<typeof group>[]) {
  return create(GithubStatusMappingSchema, {
    source: GithubStatusSource.LABEL,
    groups,
  });
}

/** 移植自 合并前的实现。 */
describe("状态映射的校验", () => {
  it("没有映射、分组过多、来源未指定各有自己的原因码", () => {
    expect(checkMapping(undefined)).toBe("MAPPING_REQUIRED");
    expect(
      checkMapping(
        labelMapping(
          ...Array.from({ length: 33 }, (_, index) =>
            group(`g${index}`, `l${index}`),
          ),
        ),
      ),
    ).toBe("TOO_MANY_GROUPS");
    expect(
      checkMapping(create(GithubStatusMappingSchema, { groups: [] })),
    ).toBe("SOURCE_UNSPECIFIED");
  });

  it("NONE 不允许带分组或 project 标识", () => {
    expect(
      checkMapping(
        create(GithubStatusMappingSchema, {
          source: GithubStatusSource.NONE,
          groups: [group("a", "todo")],
        }),
      ),
    ).toBe("SOURCE_NONE_HAS_GROUPS");
    expect(
      checkMapping(
        create(GithubStatusMappingSchema, {
          source: GithubStatusSource.NONE,
          projectId: "P_1",
        }),
      ),
    ).toBe("SOURCE_NONE_HAS_PROJECT");
    expect(
      checkMapping(
        create(GithubStatusMappingSchema, { source: GithubStatusSource.NONE }),
      ),
    ).toBe("");
  });

  it("一个仓库只有一个主来源", () => {
    const mapping = labelMapping(group("a", "todo"));
    mapping.projectId = "P_1";
    mapping.projectFieldId = "F_1";
    expect(checkMapping(mapping)).toBe("LABEL_SOURCE_HAS_PROJECT");
    expect(
      checkMapping(
        create(GithubStatusMappingSchema, {
          source: GithubStatusSource.PROJECT_FIELD,
          groups: [group("a", "")],
        }),
      ),
    ).toBe("PROJECT_IDS_REQUIRED");
  });

  it("两个分组认领同一个标签是配置错误，不是一次冲突", () => {
    expect(
      checkMapping(labelMapping(group("a", "todo"), group("b", "todo"))),
    ).toBe("GROUP_LABEL_DUPLICATE");
    expect(
      checkMapping(labelMapping(group("a", "todo"), group("a", "doing"))),
    ).toBe("GROUP_ID_INVALID");
  });

  it("耦合必须指向已知分组，而且每个状态只能耦合一次", () => {
    const mapping = labelMapping(group("a", "todo"), group("b", "done"));
    mapping.stateGroups = [
      create(GithubStateCouplingSchema, {
        state: GithubIssueState.CLOSED,
        groupId: "missing",
      }),
    ];
    expect(checkMapping(mapping)).toBe("COUPLING_GROUP_UNKNOWN");
    mapping.stateGroups = [
      create(GithubStateCouplingSchema, {
        state: GithubIssueState.CLOSED,
        groupId: "b",
      }),
      create(GithubStateCouplingSchema, {
        state: GithubIssueState.CLOSED,
        groupId: "a",
      }),
    ];
    expect(checkMapping(mapping)).toBe("COUPLING_STATE_DUPLICATE");
    mapping.stateGroups = [
      create(GithubStateCouplingSchema, {
        state: GithubIssueState.UNSPECIFIED,
        groupId: "b",
      }),
    ];
    expect(checkMapping(mapping)).toBe("COUPLING_STATE_INVALID");
  });

  it("自反的耦合是不动点，更长的环被拒", () => {
    // a 进入时关闭 → 关闭映射回 a：合法的不动点。
    const fixed = labelMapping(
      group("a", "todo", GithubIssueState.CLOSED),
      group("b", "done"),
    );
    fixed.stateGroups = [
      create(GithubStateCouplingSchema, {
        state: GithubIssueState.CLOSED,
        groupId: "a",
      }),
    ];
    expect(checkMapping(fixed)).toBe("");

    // a 进入时关闭 → 关闭映射到 b，b 进入时打开 → 打开映射回 a：一次移动会
    // 对着远端永远来回震荡。
    const cycle = labelMapping(
      group("a", "todo", GithubIssueState.CLOSED),
      group("b", "done", GithubIssueState.OPEN),
    );
    cycle.stateGroups = [
      create(GithubStateCouplingSchema, {
        state: GithubIssueState.CLOSED,
        groupId: "b",
      }),
      create(GithubStateCouplingSchema, {
        state: GithubIssueState.OPEN,
        groupId: "a",
      }),
    ];
    expect(checkMapping(cycle)).toBe("MAPPING_CYCLE");
  });
});

describe("给 Issue 标注分组", () => {
  function issue(...labels: string[]) {
    return create(GithubIssueSchema, {
      number: 1n,
      labels: labels.map((name) => create(GithubLabelSchema, { name })),
    });
  }

  it("匹配到多个标签报成冲突，而不是挑一个赢家", () => {
    const mapping = labelMapping(group("a", "todo"), group("b", "doing"));
    const both = issue("todo", "doing");
    const one = issue("doing", "unrelated");
    applyLabelGroups(mapping, [both, one]);
    expect(both.statusConflict).toBe(true);
    expect(both.statusGroupId).toBe("");
    expect(one.statusConflict).toBe(false);
    expect(one.statusGroupId).toBe("b");
  });

  it("不在 project 上的 Issue 保持未映射，而不是被猜进一个分组", () => {
    const mapping = create(GithubStatusMappingSchema, {
      source: GithubStatusSource.PROJECT_FIELD,
      projectId: "P_1",
      projectFieldId: "F_1",
      groups: [
        create(GithubStatusGroupSchema, {
          id: "a",
          title: "a",
          projectOptionId: "opt-1",
        }),
      ],
    });
    const mapped = create(GithubIssueSchema, { number: 7n });
    const absent = create(GithubIssueSchema, { number: 8n });
    applyProjectGroups(mapping, [mapped, absent], new Map([[7, "opt-1"]]));
    expect(mapped.statusGroupId).toBe("a");
    expect(absent.statusGroupId).toBe("");
  });
});
